### ISSUE TITLE
ParametricCurve, member initialization and GetLength, and GetTotalLength

### ISSUE THREAD

**Issue by bongvisu** [davideberly/GeometricTools#11]
ParametricCurve(Real tmin, Real tmax)
            :
            mTime(2),
            mSegmentLength(1),
            mAccumulatedLength(1),
            mRombergOrder(DEFAULT_ROMBERG_ORDER),
            mMaxBisections(DEFAULT_MAX_BISECTIONS),
            mConstructed(false)
        {
            mTime[0] = tmin;
            mTime[1] = tmax;
            **mSegmentLength[0] = (Real)0;
            mAccumulatedLength[0] = (Real)0;**
        }

 ParametricCurve(int numSegments, Real const* times)
            :
            mTime(numSegments + 1),
            mSegmentLength(numSegments),
            mAccumulatedLength(numSegments),
            mRombergOrder(DEFAULT_ROMBERG_ORDER),
            mMaxBisections(DEFAULT_MAX_BISECTIONS),
            mConstructed(false)
        {
            std::copy(times, times + numSegments + 1, mTime.begin());
            **mSegmentLength[0] = (Real)0;
            mAccumulatedLength[0] = (Real)0;** // maybe initializing the whole items to zeros is better?
        }
  Just the first elements are initialized to zero.
 
Real GetLength(Real t0, Real t1) const
        {
            std::function<Real(Real)> speed = [this](Real t)
            {
                return GetSpeed(t);
            };

            if (mSegmentLength[0] == (Real)0) **// checks for the first item**
            {

Real GetTotalLength() const
        {
            if (mAccumulatedLength.back() == (Real)0) **// checks for the last item, may be some undefined value**
            {

To be consistent, checking for the last item looks better. ^^
Thanks for the great code as always!!!

**Comment by davideberly** [comment]
Yes, that is a bug.

The class NaturalSplineCurve derives from ParametricCurve using the constructor that allocates mAccumulatedLength. The unit test for GetTotalLength() passes in a Debug build using Microsoft Visual Studio. With the bug, I expect the test to pass in Debug because mAccumulatedLength[] elements are all set to 0 internally (by the std::vector constructor).

However, I would have expected the test to fail in Release assuming that the elements of mAccumulatedLength[] are uninitialized. For whatever reason, the MSVS code is initializing the elements to 0 in the Release build, and the unit test passes. In Release, I set a breakpoint at the test “if (mAccumulatedLength.back() == (Real)0)” and executed the test with a debugger attached. The breakpoint is tagged with a comment indicating that the line of code was optimized away, which is in my opinion an error. In this case, the GetLength(…) function is called and has the side effect of assigning all elements of mAccumulatedLength[]. I then modified the Release build to disable optimizations, but once again the construction of mAccumulatedLength[] sets all elements to zero (internally in std::vector constructor). I ran the test through Intel Inspector, but it has no complaints about reading uninitialized memory--as one would expect in the first line of GetTotalLength() if mAccumulatedLength[] has uninitialized elements.

If the constructor call is set to “mAccumulatedLength(numSegments, std::numeric_limits<Real>::max())”, the GetLength(…) function is skipped and GetTotalLength() returns std::numeric_limits<Real>::max(), which demonstrates the bug.

As you mention, a simple fix is to initialize all the elements explicitly in the std::vector (rather than just the “back” one). I will post the fix shortly.

Thank you for taking the time to report the bug.

--
Dave Eberly
https://www.geometrictools.com
